File: Leetcode Daily Challenge/025-July-2025/028-count-number-of-maximum-bitwise-or-subsets.java
// https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/




import java.util.*;

class Solution {
    // TC : O(N * maxOr)
    // SC : O(N * maxOr)
    public int countMaxOrSubsets(int[] nums) {
        int maxOr = 0;
        for (int num : nums) maxOr |= num;

        int[][] dp = new int[nums.length + 1][maxOr + 1];
        for (int i = 0; i < dp.length; i ++) {
            Arrays.fill(dp[i], -1);
        }
        return solve(0, 0, nums, maxOr, dp);
    }
    private int solve(int i, int currOr, int[] nums, int maxOr, int[][] dp) {
        if (i >= nums.length) {
            if (currOr == maxOr)
                return 1;
            return 0;
        }

        if (dp[i][currOr] != -1) return dp[i][currOr];

        int op1 = solve(i + 1, currOr | nums[i], nums, maxOr, dp);
        int op2 = solve(i + 1, currOr, nums, maxOr, dp);

        dp[i][currOr] = (op1 + op2);
        return dp[i][currOr];
    }
}

File: Leetcode Daily Challenge/025-July-2025/028-count-number-of-maximum-bitwise-or-subsets.cpp
// https://leetcode.com/problems/count-number-of-maximum-bitwise-or-subsets/




#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    // TC : O(N * maxOr)
    // SC : O(N * maxOr)
    int solve(int idx, vector<int>& nums, int sum, int maxOr, vector<vector<int>>& dp) {
        if (idx >= nums.size()) {
            if (sum == maxOr) 
                return 1;
            return 0;
        }

        if (dp[idx][sum] != -1) return dp[idx][sum];

        int take    = solve(idx + 1, nums, sum | nums[idx], maxOr, dp);        
        int notTake = solve(idx + 1, nums, sum, maxOr, dp);

        return dp[idx][sum] = (take + notTake);
    }
public:
    int countMaxOrSubsets(vector<int>& nums) {
        int maxOr = 0;
        for (int &num : nums) {
            maxOr |= num;
        }

        int sum = 0;
        vector<vector<int>> dp(nums.size() + 1, vector<int>(maxOr + 1, -1));
        return solve(0, nums, sum, maxOr, dp);
    }
};